Added more accurate version of the `betaln` function.

imh opened the pull request:
Resolves #13267 

I'm not sure about it style-guide-wise, and I'm just pattern-matching on how to use `_wraps`, but this version is much more accurate than the current.

I was unable to get bazel working and haven't run the tests locally, but did compare accuracy against scipy and the current implementation (see the issue) on CPU.

imh committed:
Added more accurate version of the betaln function.

jakevdp commented on jax/_src/scipy/special.py:
```diff
@@ -38,10 +39,7 @@ def gammaln(x: ArrayLike) -> Array:
-def betaln(x: ArrayLike, y: ArrayLike) -> Array:
-  x, y = _promote_args_inexact("betaln", x, y)
-  return lax.lgamma(x) + lax.lgamma(y) - lax.lgamma(x + y)
+_wraps(betaln, module='scipy.special', update_doc=False)
```
this should be something like
```python
betaln = _wraps(osp_special.betaln, module='scipy.special')(betaln)
```
As currently written, the output of `_wraps` is unassigned and so it will have no effect.

imh replied:
Great catch, thanks

jakevdp reviewed (changes requested):
Looks great, thanks! Could you add some tests that cover the problematic cases?

imh commented:
Thanks for the feedback. I added a test case testing `betaln(x, 1)` and its derivative across a wide range of x, and I think I fixed the wrapper.

A few questions:

- `betaln` doesn't currently show up in [the API docs](https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.special). I've checked that this change keeps it that way when locally rebuilding the docs. Is that the intended case, or is it intended to show up?
- I have only tested on CPU. Does the automated test suite run the tests on GPU/TPU? If not, I can spin some up and test myself.

hawkinsp commented:
Incidentally:

> I was unable to get bazel working and haven't run the tests locally

You don't need bazel to run the tests. You can use either `pytest` or `bazel`. See: https://jax.readthedocs.io/en/latest/developer.html#using-pytest

(We use `bazel` to build the C++ parts of JAX, but if you're changing the Python parts only you don't need to build anything and can use a prebuilt `jaxlib` and run tests with `pytest`, etc.)

> betaln doesn't currently show up in [the API docs](https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.special).

It's probably missing from the doc index. To make it show up, add it to this doc: https://github.com/google/jax/blob/main/docs/jax.scipy.rst

> I have only tested on CPU. Does the automated test suite run the tests on GPU/TPU?

Yes. A GPU test runs as a Github presubmit, and we'll also run TPU tests before merging.

jakevdp commented on docs/jax.scipy.rst:
```diff
@@ -106,6 +106,7 @@ jax.scipy.special
 .. autosummary::
   :toctree: _autosummary
 
+   betaln
```
Move down one line for alphabetization

imh replied:
WOW. It wasn't that early in the morning that I couldn't do my ABC's ... or so I thought. 🙏

jakevdp reviewed (changes requested):
Thanks! A couple minor suggestions. Additionally, please squash your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests)

jakevdp commented:
Also if you rebase on the main branch the readthedocs issue should resolve itself.

imh commented:
Thank you again. I made the requested changes, and I think we should be ready to go now.

jakevdp reviewed (approved):
Looks great, thanks!

jakevdp commented on tests/lax_scipy_test.py:
```diff
@@ -303,6 +303,17 @@ def testIssue3758(self):
+    primals_out, tangents_out = jax.jvp(lsp_special.betaln, primals=[xs, 1.0], tangents=[jnp.ones_like(xs), 0.0])
+    # Check that betaln(x, 1) = -log(x).
+    # Betaln is still not perfect for small values, hence the atol (but it's close)
+    assert jnp.allclose(primals_out, -jnp.log(xs), atol=1e-6)
```
It looks like this is failing on TPU; I'd probably just skip this test on TPU.

imh commented:
Out of an abundance of caution, since it's failing on TPU, I spun up a TPU and did some checks.

As expected, the version currently in JAX is inaccurate on TPU and CPU:
![Old_betaln_TPU](https://user-images.githubusercontent.com/391217/204434408-8c3c1e55-6f02-46ef-879e-62fe1e6e4ee7.png)
![Old_betaln_CPU](https://user-images.githubusercontent.com/391217/204434414-761d78fc-12c0-44aa-bd8b-2508b1795f03.png)

The new version in this PR isn't as accurate on TPU as it is on CPU, but it's tolerable on both:
![New_betaln_TPU](https://user-images.githubusercontent.com/391217/204434416-eca7328b-6541-4b6f-a0f4-fb40bd7d37a2.png)
![New_betaln_CPU](https://user-images.githubusercontent.com/391217/204434418-98c91475-6cce-4675-93e4-425bf3c832bb.png)

I'll come back to this tomorrow to disable the test on TPU.

imh commented:
FYI, after you approved it, I made these changes:
- Added a specific error tolerance for TPU, so the tests should be working on them now
- Changed `assert jnp.allclose(...)` to `self.assertAllClose(...)`

jakevdp reviewed (approved):
Looks good, thanks!

jakevdp commented:
It looks like there are still some failures on CPU and GPU - are you able to see them under the checks below?

imh commented:
I can see the invocation log for the GPU build, but not the CPU build ("Unable to retrieve invocation log."). I can't download the `build.log` artifact for either ("Error 403, Forbidden")

- It looks like the GPU test just needs some looser tolerances, which I've just pushed (1e-6 -> 1e-5)
- If you can see the log for the CPU build, what's the failure there?

jakevdp commented:
We'd expect much tighter tolerance for float64 right? How about using a pattern like this to use appropriate tolerance depending on the dtype? https://github.com/google/jax/blob/7495a9e3705d67c498723240cfed3f14c5916567/tests/lax_scipy_test.py#L206

imh commented:
The float64 error is dominated by the function approximation error rather than floating point error, so it only cuts the absolute error in half:
![New_betaln_GPU_float64 (1)](https://user-images.githubusercontent.com/391217/204643658-93ece274-447e-49c5-a3f9-eab6ec59de3d.png)

What was the CPU error? The tests pass locally, so I'm sorry to be making you hit the button again and again.

jakevdp commented:
Thanks for looking into that! I couldn't see the CPU error logs either. Might have been a flake.
